connmgr/connmanager_test.go

```diff
@@ -0,0 +1,373 @@
+func (c mockConn) SetReadDeadline(t time.Time) error  { return nil }
+func (c mockConn) SetWriteDeadline(t time.Time) error { return nil }
+
+// mockDialer mocks the net.Dial interface by returning a mock connection to
+// the given address.
+func mockDialer(network, address string) (net.Conn, error) {
+	r, w := io.Pipe()
+	c := &mockConn{raddr: address}
+	c.Reader = r
+	c.Writer = w
+	return c, nil
+}
+
+// TestNewConfig tests that new ConnManager config is validated as expected.
+func TestNewConfig(t *testing.T) {
+	_, err := connmgr.New(&connmgr.Config{})
+	if err == nil {
+		t.Fatalf("connmgr.New expected error: 'Dial can't be nil', got nil")
+	}
+	_, err = connmgr.New(&connmgr.Config{
+		Dial: mockDialer,
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New unexpected error: %v", err)
+	}
+}
+
+// TestUseLogger tests that a logger can be passed to UseLogger
+func TestUseLogger(t *testing.T) {
+	l, err := btclog.NewLoggerFromWriter(bytes.NewBuffer(nil), btclog.InfoLvl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	connmgr.UseLogger(l)
+}
+
+// TestStartStop tests that the connection manager starts and stops as
+// expected.
+func TestStartStop(t *testing.T) {
+	connected := make(chan *connmgr.ConnReq)
+	cmgr, err := connmgr.New(&connmgr.Config{
+		MaxOutbound:   1,
+		GetNewAddress: func() string { return "127.0.0.1:18555" },
+		Dial:          mockDialer,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			connected <- c
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+	cmgr.Start()
+	go cmgr.Wait()
+	gotConnReq := <-connected
+	cmgr.Stop()
+	// already stopped
+	cmgr.Stop()
+	// ignored
+	cr := &connmgr.ConnReq{Addr: "127.0.0.1:18555", Permanent: true}
+	cmgr.Connect(cr)
+	cmgr.Disconnect(gotConnReq.ID())
+	cmgr.Remove(gotConnReq.ID())
+}
+
+// TestConnectMode tests that the connection manager works in the connect mode.
+//
+// In connect mode, automatic connections are disabled, so we test that
+// requests using Connect are handled and that no other connections are made.
+func TestConnectMode(t *testing.T) {
+	connected := make(chan *connmgr.ConnReq)
+	cmgr, err := connmgr.New(&connmgr.Config{
+		MaxOutbound: 2,
+		Dial:        mockDialer,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			connected <- c
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+	cr := &connmgr.ConnReq{Addr: "127.0.0.1:18555", Permanent: true}
+	cmgr.Start()
+	cmgr.Connect(cr)
+	gotConnReq := <-connected
+	wantID := cr.ID()
+	gotID := gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("connect mode: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState := cr.State()
+	wantState := connmgr.ConnEstablished
+	if gotState != wantState {
+		t.Fatalf("connect mode: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+	select {
+	case c := <-connected:
+		t.Fatalf("connect mode: got unexpected connection - %v", c.Addr)
+	case <-time.After(time.Millisecond):
```

Comment:
Tested by forcing a mock connection in `Start` and checked that the test failed. Since we're using `mockDial` I think it is fast enough.